# Project 3

In this project, you will perform a logistic regression on the admissions data we've been working with in projects 1 and 2.

In [301]:
%matplotlib inline
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.api as sm
import pylab as pl
import numpy as np


In [302]:
df_raw = pd.read_csv("/Users/vvnikitin/Desktop/admissions.csv")
df = df_raw.dropna() 
print df.head()

   admit    gre   gpa  prestige
0      0  380.0  3.61       3.0
1      1  660.0  3.67       3.0
2      1  800.0  4.00       1.0
3      1  640.0  3.19       4.0
4      0  520.0  2.93       4.0


## Part 1. Frequency Tables

#### 1. Let's create a frequency table of our variables

In [304]:
# frequency table for prestige and whether or not someone was admitted
my_tab = pd.crosstab(index=df['admit'], columns=df['prestige'])
my_tab

prestige  1.0  2.0  3.0  4.0
admit                       
0          28   95   93   55
1          33   53   28   12

## Part 2. Return of dummy variables

#### 2.1 Create class or dummy variables for prestige 

In [305]:
dummy_ranks = pd.get_dummies(df['prestige'], prefix='prestige')
print dummy_ranks.head()

   prestige_1.0  prestige_2.0  prestige_3.0  prestige_4.0
0             0             0             1             0
1             0             0             1             0
2             1             0             0             0
3             0             0             0             1
4             0             0             0             1


#### 2.2 When modeling our class variables, how many do we need? 



Answer: we need 3(three)

## Part 3. Hand calculating odds ratios

Develop your intuition about expected outcomes by hand calculating odds ratios.

In [41]:
cols_to_keep = ['admit', 'gre', 'gpa']
handCalc = df[cols_to_keep].join(dummy_ranks.loc[:, 'prestige':])
print handCalc.head()

   admit    gre   gpa  prestige_1.0  prestige_2.0  prestige_3.0  prestige_4.0
0      0  380.0  3.61             0             0             1             0
1      1  660.0  3.67             0             0             1             0
2      1  800.0  4.00             1             0             0             0
3      1  640.0  3.19             0             0             0             1
4      0  520.0  2.93             0             0             0             1


In [46]:
#crosstab prestige 1 admission 
# frequency table cutting prestige and whether or not someone was admitted
my_tab_1 = pd.crosstab(index=handCalc['prestige_1.0'], columns=df['admit'])
my_tab_1

admit           0   1
prestige_1.0         
0             243  93
1              28  33

#### 3.1 Use the cross tab above to calculate the odds of being admitted to grad school if you attended a #1 ranked college

In [102]:
WA = 33
LA = 28
odds1 = float(WA)/LA
print 'odds1 = %.2f' % odds1

odds1 = 1.18


#### 3.2 Now calculate the odds of admission if you did not attend a #1 ranked college

In [96]:
WB = 93
LB = 243
odds2 = float(WB)/LB
print 'odds2 = %.2f' % odds2

odds2 = 0.383


#### 3.3 Calculate the odds ratio

In [103]:
odds_ratio = odds1/odds2
print 'odds ratio = %.2f' % odds_ratio

odds ratio = 3.08


#### 3.4 Write this finding in a sentenance: 

Answer: OR = 3.08 >1 Exposure associated with higher odds of outcome

#### 3.5 Print the cross tab for prestige_4

In [67]:
my_tab_4 = pd.crosstab(index=handCalc['prestige_4.0'], columns=df['admit'])
my_tab_4

admit           0    1
prestige_4.0          
0             216  114
1              55   12

#### 3.6 Calculate the OR 

In [116]:
ad = 12.0*216.0
bc = 55.0*114.0
odds_ratio = ad/bc
print 'odds ratio = %.2f' % odds_ratio

odds ratio = 0.41


#### 3.7 Write this finding in a sentence

Answer: OR<1 Exposure associated with lower odds of outcome

## Part 4. Analysis

In [306]:
# create a clean data frame for the regression
cols_to_keep = ['admit', 'gre', 'gpa']
data = df[cols_to_keep].join(dummy_ranks.loc[:, 'prestige_2.0':])
print data.head()

   admit    gre   gpa  prestige_2.0  prestige_3.0  prestige_4.0
0      0  380.0  3.61             0             1             0
1      1  660.0  3.67             0             1             0
2      1  800.0  4.00             0             0             0
3      1  640.0  3.19             0             0             1
4      0  520.0  2.93             0             0             1


We're going to add a constant term for our Logistic Regression. The statsmodels function we're going to be using requires that intercepts/constants are specified explicitly.

In [307]:
# manually add the intercept
data['intercept'] = 1.0

In [308]:
print train_cols

Index([u'gre', u'gpa', u'prestige_2.0', u'prestige_3.0', u'prestige_4.0',
       u'intercept'],
      dtype='object')


#### 4.1 Set the covariates to a variable called train_cols

In [309]:
train_cols = data.columns[1:]


Optimization terminated successfully.
         Current function value: 0.573854
         Iterations 6


In [310]:
data[train_cols].head()

gre   gpa  prestige_2.0  prestige_3.0  prestige_4.0  intercept
0  380.0  3.61             0             1             0        1.0
1  660.0  3.67             0             1             0        1.0
2  800.0  4.00             0             0             0        1.0
3  640.0  3.19             0             0             1        1.0
4  520.0  2.93             0             0             1        1.0

#### 4.2 Fit the model

In [312]:
logit = sm.Logit(data['admit'], data[train_cols])
result = logit.fit()

Optimization terminated successfully.
         Current function value: 0.573854
         Iterations 6


#### 4.3 Print the summary results

In [313]:
print result.summary()

                           Logit Regression Results                           
Dep. Variable:                  admit   No. Observations:                  397
Model:                          Logit   Df Residuals:                      391
Method:                           MLE   Df Model:                            5
Date:                Wed, 06 Dec 2017   Pseudo R-squ.:                 0.08166
Time:                        18:59:16   Log-Likelihood:                -227.82
converged:                       True   LL-Null:                       -248.08
                                        LLR p-value:                 1.176e-07
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
gre              0.0022      0.001      2.028      0.043    7.44e-05       0.004
gpa              0.7793      0.333      2.344      0.019       0.128       1.431
prestige_2.0    -0.6801      0.317     -2.14

#### 4.4 Calculate the odds ratios of the coeffiencents and their 95% CI intervals

hint 1: np.exp(X)

hint 2: conf['OR'] = params
        
           conf.columns = ['2.5%', '97.5%', 'OR']

In [314]:
params = result.params
conf = result.conf_int()
conf['OR'] = params
conf.columns = ['2.5%', '97.5%', 'OR']
print np.exp(conf)

                  2.5%     97.5%        OR
gre           1.000074  1.004372  1.002221
gpa           1.136120  4.183113  2.180027
prestige_2.0  0.272168  0.942767  0.506548
prestige_3.0  0.133377  0.515419  0.262192
prestige_4.0  0.093329  0.479411  0.211525
intercept     0.002207  0.194440  0.020716


#### 4.5 Interpret the OR of Prestige_2

Answer: OR of Prestige-2 is 0.38 < 1,so exposure associated with lower odds of outcome

#### 4.6 Interpret the OR of GPA

Answer: OR of GPA is 0.87 < 1, so exposure associated with lower odds of outcome

## Part 5: Predicted probablities


As a way of evaluating our classifier, we're going to recreate the dataset with every logical combination of input values. This will allow us to see how the predicted probability of admission increases/decreases across different variables. First we're going to generate the combinations using a helper function called cartesian (above).

We're going to use np.linspace to create a range of values for "gre" and "gpa". This creates a range of linearly spaced values from a specified min and maximum value--in our case just the min/max observed values.

In [315]:
def cartesian(arrays, out=None):
    """
    Generate a cartesian product of input arrays.
    Parameters
    ----------
    arrays : list of array-like
        1-D arrays to form the cartesian product of.
    out : ndarray
        Array to place the cartesian product in.
    Returns
    -------
    out : ndarray
        2-D array of shape (M, len(arrays)) containing cartesian products
        formed of input arrays.
    Examples
    --------
    >>> cartesian(([1, 2, 3], [4, 5], [6, 7]))
    array([[1, 4, 6],
           [1, 4, 7],
           [1, 5, 6],
           [1, 5, 7],
           [2, 4, 6],
           [2, 4, 7],
           [2, 5, 6],
           [2, 5, 7],
           [3, 4, 6],
           [3, 4, 7],
           [3, 5, 6],
           [3, 5, 7]])
    """

    arrays = [np.asarray(x) for x in arrays]
    dtype = arrays[0].dtype

    n = np.prod([x.size for x in arrays])
    if out is None:
        out = np.zeros([n, len(arrays)], dtype=dtype)

    m = n / arrays[0].size
    out[:,0] = np.repeat(arrays[0], m)
    if arrays[1:]:
        cartesian(arrays[1:], out=out[0:m,1:])
        for j in xrange(1, arrays[0].size):
            out[j*m:(j+1)*m,1:] = out[0:m,1:]
    return out

In [316]:
# instead of generating all possible values of GRE and GPA, we're going
# to use an evenly spaced range of 10 values from the min to the max 
gres = np.linspace(data['gre'].min(), data['gre'].max(), 10)
print gres
# array([ 220.        ,  284.44444444,  348.88888889,  413.33333333,
#         477.77777778,  542.22222222,  606.66666667,  671.11111111,
#         735.55555556,  800.        ])
gpas = np.linspace(data['gpa'].min(), data['gpa'].max(), 10)
print gpas
# array([ 2.26      ,  2.45333333,  2.64666667,  2.84      ,  3.03333333,
#         3.22666667,  3.42      ,  3.61333333,  3.80666667,  4.        ])


# enumerate all possibilities
combos = pd.DataFrame(cartesian([gres, gpas, [1, 2, 3, 4], [1.]]))

[ 220.          284.44444444  348.88888889  413.33333333  477.77777778
  542.22222222  606.66666667  671.11111111  735.55555556  800.        ]
[ 2.26        2.45333333  2.64666667  2.84        3.03333333  3.22666667
  3.42        3.61333333  3.80666667  4.        ]


#### 5.1 Recreate the dummy variables

In [317]:
# recreate the dummy variables

combos.columns = ['gre', 'gpa', 'prestige', 'intercept']
dummy_ranks = pd.get_dummies(combos['prestige'], prefix='prestige')
dummy_ranks.columns = ['prestige_1.0', 'prestige_2.0', 'prestige_3.0', 'prestige_4.0']

# keep only what we need for making prediction

cols_to_keep = ['gre', 'gpa', 'intercept']
combos = combos[cols_to_keep].join(dummy_ranks.ix[:, 'prestige_2':])

/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:10: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  # Remove the CWD from sys.path while we load stuff.


#### 5.2 Make predictions on the enumerated dataset

In [319]:
combos = combos[train_cols]
combos.tail()

gre       gpa  prestige_2.0  prestige_3.0  prestige_4.0  intercept
395  800.0  3.806667             0             0             1        1.0
396  800.0  4.000000             0             0             0        1.0
397  800.0  4.000000             1             0             0        1.0
398  800.0  4.000000             0             1             0        1.0
399  800.0  4.000000             0             0             1        1.0

In [320]:
combos['admit_pred'] = result.predict(combos)
print combos.tail()

       gre       gpa  prestige_2.0  prestige_3.0  prestige_4.0  intercept  \
395  800.0  3.806667             0             0             1        1.0   
396  800.0  4.000000             0             0             0        1.0   
397  800.0  4.000000             1             0             0        1.0   
398  800.0  4.000000             0             1             0        1.0   
399  800.0  4.000000             0             0             1        1.0   

     admit_pred  
395    0.334286  
396    0.734040  
397    0.582995  
398    0.419833  
399    0.368608  


#### 5.3 Interpret findings for the last 4 observations

Answer: The probability of admission gradually increases as gre and gpa increase and that the different prestige levels yield drastic probabilities of admission (particularly the most prestigious schools).

## Bonus

Plot the probability of being admitted into graduate school, stratified by GPA and GRE score.